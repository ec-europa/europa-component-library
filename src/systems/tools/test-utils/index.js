const merge = require('deepmerge');
const twing = require('@ecl-twig/ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const test = document.createElement('test');
      test.innerHTML = html.trim();
      resolve(test);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
