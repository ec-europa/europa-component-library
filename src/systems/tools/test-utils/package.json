{
  "name": "@ecl-twig/test-utils",
  "version": "2.31.1",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
