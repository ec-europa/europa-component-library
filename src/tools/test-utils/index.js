const merge = require('deepmerge');
const twing = require('@ecl/twig-ec-storybook/.storybook/environment');

const renderTwigFileAsNode = (file, options) =>
  new Promise((resolve, reject) => {
    try {
      const html = twing.render(file, options);
      const div = document.createElement('div');
      div.innerHTML = html.trim();
      resolve(div.firstChild);
    } catch (error) {
      reject(error);
    }
  });

module.exports = {
  merge,
  renderTwigFileAsNode,
};
