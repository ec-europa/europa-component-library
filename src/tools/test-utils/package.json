{
  "name": "@ecl/test-utils",
  "version": "3.4.2",
  "description": "A local test helper",
  "main": "index.js",
  "private": true,
  "devDependencies": {
    "deepmerge": "4.2.2"
  }
}
