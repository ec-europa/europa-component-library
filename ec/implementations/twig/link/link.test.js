import { merge, renderTwigFileAsNode } from '@ecl-twig/test-utils';
import dataDefault from './demo/data--default';
import dataCta from './demo/data--call-to-action';
import dataStandalone from './demo/data--standalone';

describe('EC - Link', () => {
  const template = '@ecl-twig/ec-component-link/ecl-link.html.twig';
  const render = (params) => renderTwigFileAsNode(template, params);
  const defaultIconPath = 'static/icons.svg';

  describe('Default', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataDefault)).resolves.toMatchSnapshot();
    });
  });

  describe('Standalone', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(dataStandalone)).resolves.toMatchSnapshot();
    });
  });

  describe('Standalone with missing icon name', () => {
    const options = merge(dataStandalone, {
      link: {
        label: 'Standalone link with icon',
        icon_position: 'after',
      },
      icon: {
        type: 'ui',
        size: 'fluid',
        path: defaultIconPath,
      },
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('With two icons after', () => {
    const options = merge(dataStandalone, {
      link: {
        type: 'standalone',
        label: 'Standalone link with icon',
        icon_position: 'after',
      },
      icon: [
        {
          type: 'ui',
          name: 'external',
          size: 'fluid',
          path: defaultIconPath,
        },
        {
          type: 'ui',
          name: 'download',
          size: 'fluid',
          path: defaultIconPath,
        },
      ],
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(dataStandalone, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });
  });

  describe('With two icons before', () => {
    const options = merge(dataStandalone, {
      link: {
        type: 'standalone',
        label: 'Standalone link with icon',
        icon_position: 'before',
      },
      icon: [
        {
          type: 'ui',
          name: 'external',
          size: 'fluid',
          path: defaultIconPath,
        },
        {
          type: 'ui',
          name: 'download',
          size: 'fluid',
          path: defaultIconPath,
        },
      ],
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(options, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });
  });

  describe('With icon before', () => {
    test('renders correctly', () => {
      expect.assertions(1);

      const options = merge(dataStandalone, {
        link: {
          label: 'Standalone link with icon',
          icon_position: 'before',
        },
        icon: {
          type: 'ui',
          name: 'external',
          size: 'fluid',
          path: defaultIconPath,
          extra_classes: 'ecl-test-extra-class',
        },
      });

      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('With icon after', () => {
    const options = merge(dataStandalone, {
      link: {
        label: 'Standalone link with icon',
        icon_position: 'after',
      },
      icon: {
        type: 'ui',
        name: 'external',
        size: 'fluid',
        path: defaultIconPath,
        extra_classes: 'ecl-test-extra-class',
      },
    });

    test('renders correctly', () => {
      expect.assertions(1);
      return expect(render(options)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra class names', () => {
      expect.assertions(1);

      const withExtraClasses = merge(options, {
        extra_classes: 'custom-class custom-class--test',
      });

      return expect(render(withExtraClasses)).resolves.toMatchSnapshot();
    });

    test('renders correctly with extra attributes', () => {
      expect.assertions(1);

      const withExtraAttributes = merge(options, {
        extra_attributes: [
          { name: 'data-test', value: 'data-test-value' },
          { name: 'data-test-1', value: 'data-test-value-1' },
        ],
      });

      return expect(render(withExtraAttributes)).resolves.toMatchSnapshot();
    });
  });

  describe('Cta variant with icon', () => {
    const options = merge(dataCta, {
      link: {
        label: 'Call to action link',
        icon_position: 'after',
      },
      icon: {
        type: 'ui',
        name: 'corner-arrow',
        size: 'fluid',
        path: defaultIconPath,
        transform: 'rotate-90',
      },
    });

    test('renders correctly', () => {
      expect.assertions(1);

      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });

  describe('with validation enabled and missing input data', () => {
    test('returns the right warning message', () => {
      expect.assertions(1);

      const options = {
        link: {
          label: '',
          path: '',
          type: 'standalone',
          icon_position: 'after',
        },
        icon: {
          name: 'check-filled',
        },
        _compliance_: true,
      };

      return expect(render(options)).resolves.toMatchSnapshot();
    });
  });
});
