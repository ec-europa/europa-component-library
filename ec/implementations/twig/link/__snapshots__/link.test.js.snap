// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EC - Link Cta variant with icon renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the CTA link component"
    class="ecl-link ecl-link--cta ecl-link--icon ecl-link--icon-after"
    href="/example#link-cta"
  >
    <span
      class="ecl-link__label"
    >
      Call to action link
    </span>
     
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-icon--rotate-90 ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--corner-arrow"
      />
    </svg>
  </a>
</test>
`;

exports[`EC - Link Default renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the European Union"
    class="ecl-link ecl-link--default"
    href="/example#link-default"
  >
    The European Union
  </a>
</test>
`;

exports[`EC - Link Standalone renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone"
    href="/example#link-standalone"
  >
    Standalone link
  </a>
</test>
`;

exports[`EC - Link Standalone with missing icon name renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone"
    href="/example#link-standalone"
  >
    Standalone link with icon
  </a>
</test>
`;

exports[`EC - Link With icon after renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
    href="/example#link-standalone"
  >
    <span
      class="ecl-link__label"
    >
      Standalone link with icon
    </span>
     
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--external"
      />
    </svg>
  </a>
</test>
`;

exports[`EC - Link With icon after renders correctly with extra attributes 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
    data-test="data-test-value"
    data-test-1="data-test-value-1"
    href="/example#link-standalone"
  >
    <span
      class="ecl-link__label"
    >
      Standalone link with icon
    </span>
     
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--external"
      />
    </svg>
  </a>
</test>
`;

exports[`EC - Link With icon after renders correctly with extra class names 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after custom-class custom-class--test"
    href="/example#link-standalone"
  >
    <span
      class="ecl-link__label"
    >
      Standalone link with icon
    </span>
     
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--external"
      />
    </svg>
  </a>
</test>
`;

exports[`EC - Link With icon before renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-before"
    href="/example#link-standalone"
  >
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon ecl-test-extra-class"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--external"
      />
    </svg>
     
    <span
      class="ecl-link__label"
    >
      Standalone link with icon
    </span>
  </a>
</test>
`;

exports[`EC - Link With two icons after renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
    href="/example#link-standalone"
  >
    <span
      class="ecl-link__label"
    >
      Standalone link with icon
    </span>
     
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--external"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--download"
      />
    </svg>
  </a>
</test>
`;

exports[`EC - Link With two icons after renders correctly with extra class names 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone custom-class custom-class--test"
    href="/example#link-standalone"
  >
    Standalone link
  </a>
</test>
`;

exports[`EC - Link With two icons before renders correctly 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-before"
    href="/example#link-standalone"
  >
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--external"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--download"
      />
    </svg>
     
    <span
      class="ecl-link__label"
    >
      Standalone link with icon
    </span>
  </a>
</test>
`;

exports[`EC - Link With two icons before renders correctly with extra class names 1`] = `
<test>
  <a
    aria-label="Read more about the Standalone link"
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-before custom-class custom-class--test"
    href="/example#link-standalone"
  >
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--external"
      />
    </svg>
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--fluid ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="static/icons.svg#ui--download"
      />
    </svg>
     
    <span
      class="ecl-link__label"
    >
      Standalone link with icon
    </span>
  </a>
</test>
`;

exports[`EC - Link with validation enabled and missing input data returns the right warning message 1`] = `
<test>
  <a
    class="ecl-link ecl-link--standalone ecl-link--icon ecl-link--icon-after"
    href=""
  >
    <span
      class="ecl-link__label"
    />
     
    <svg
      aria-hidden="true"
      class="ecl-icon ecl-icon--m ecl-link__icon"
      focusable="false"
    >
      <use
        xlink:href="#check-filled"
      />
    </svg>
    <!-- ecl2-compliance: icon -->
    <div
      class="ecl2-compliance ecl-u-border-width-1 ecl-u-border-color-red ecl-u-border-style-solid ecl-u-ma-xs ecl-u-pr-s ecl-u-bg-white"
    >
      <p
        class="ecl-u-type-color-grey ecl-u-ml-m ecl-u-type-xs"
      >
        [icon] - {{ icon.name }} and {{ icon.path }} properties need to set and filled in.
      </p>
    </div>
  </a>
  <!-- ecl2-compliance: link--standalone -->
  <div
    class="ecl2-compliance ecl-u-border-width-1 ecl-u-border-color-red ecl-u-border-style-solid ecl-u-ma-xs ecl-u-pr-s ecl-u-bg-white"
  >
    <p
      class="ecl-u-type-color-grey ecl-u-ml-m ecl-u-type-xs"
    >
      [link--standalone] - {{ link.label }} and none of {{ link.icon_position }}, {{ icon.name }}, {{ icon.path }} or {{ icon.type }} can be empty at the same time, an unusable link would be generated otherwise
    </p>
  </div>
</test>
`;
